File: sample_source_code/javascript/src/Animal.js
const IAnimal = require('./IAnimal');

/**
 * Abstract base class providing shared state and describe().
 * @extends IAnimal
 */
class Animal extends IAnimal {
    /**
     * @param {string} name
     * @param {number} age
     * @param {string} species
     * @param {number} weight
     */
    constructor(name, age, species, weight) {
        super();
        this._name    = name;
        this._age     = age;
        this._species = species;
        this._weight  = weight;
    }

    getName()    { return this._name; }
    getSpecies() { return this._species; }
    getAge()     { return this._age; }
    getWeight()  { return this._weight; }

    describe() {
        return `${this._name} (${this._species}), age=${this._age}, ` +
               `weight=${this._weight.toFixed(1)}kg, sound='${this.sound()}'`;
    }

    toString() { return this.describe(); }
}

module.exports = Animal;

File: sample_source_code/CPP/src/Animal.cpp
#include "Animal.h"
#include <sstream>
#include <iomanip>

Animal::Animal(const std::string& name, int age, const std::string& species, double weight)
    : name_(name), age_(age), species_(species), weight_(weight) {}

std::string Animal::getName()    const { return name_; }
std::string Animal::getSpecies() const { return species_; }
int         Animal::getAge()     const { return age_; }
double      Animal::getWeight()  const { return weight_; }

std::string Animal::describe() const {
    std::ostringstream ss;
    ss << name_ << " (" << species_ << "), age=" << age_
       << ", weight=" << std::fixed << std::setprecision(1) << weight_ << "kg"
       << ", sound='" << sound() << "'";
    return ss.str();
}
